src/shared/sleep-config.c

```diff
@@ -237,45 +384,58 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
                                 log_debug("Ignoring compressed RAM swap device '%s'.", swap->device);
                                 continue;
                         }
+                } else {
+                        log_debug("Swap type %s is unsupported for hibernation: %s; skipping", swap->type, swap->device);
+                        continue;
                 }
 
-                /* prefer highest priority or swap with most remaining space when same priority */
-                if (!selected_swap || swap->priority > selected_swap->priority
-                    || ((swap->priority == selected_swap->priority)
-                        && (swap->size - swap->used) > (selected_swap->size - selected_swap->used))) {
-                        selected_swap = swap_entry_free(selected_swap);
-                        selected_swap = TAKE_PTR(swap);
+                /* prefer resume device or highest priority swap with most remaining space */
+                if (!hibernate_location || swap->priority > hibernate_location->swap->priority
+                    || ((swap->priority == hibernate_location->swap->priority)
+                        && (swap->size - swap->used) > (hibernate_location->swap->size - hibernate_location->swap->used))) {
+
+                        _cleanup_free_ char *swap_device_id = NULL;
+                        r = swap_device_to_major_minor(swap, &swap_device_id);
+                        if (r < 0)
+                                return r;
+
+                        hibernate_location = hibernate_location_free(hibernate_location);
+                        r = hibernate_location_new(TAKE_PTR(swap), TAKE_PTR(swap_device_id), swap_offset, &hibernate_location);
```

Comment:
If hibernate_location_new fails, this will leak both pointers. I think it'd be easier to just allocate the structure here.